### PR TITLE
Add kalpy recipe

### PR DESCRIPTION
**Pull request by mmcauliffe**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by mmcauliffe** [comment]
@h-vetinari This is for pybind11 bindings on the shared libraries of kaldi, if you could take a look over the recipe and https://github.com/mmcauliffe/kalpy, definitely appreciate any feedback! I have the low level bindings working along with some higher level pythonic classes/functions that I'll be developing out further, but wanted to get the ball rolling on distribution.

File: recipes/kalpy/bld.bat
Comment (h-vetinari) [inline recipes/kalpy/bld.bat:4]: We don't need this on windows, that's OSX-only

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:10]: Can you explain the situation a little. On the README of , I see:
> Kalpy is also available on pip via the mfa-kaldi package

what's the relationship between kalpy and mfa-kaldi? I guess kalpy is the more fully-featured one (includes the lib)? Should we add an output `mfa-kaldi` here that just wraps `kalpy` to ensure that no-one gets the idea of packaging `mfa-kalpi` separately?

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:19]: I think this is easier - less indirections.
```suggestion
   - KALDI_ROOT=%LIBRARY_PREFIX%  # [win]
   - KALDI_ROOT=$PREFIX           # [unix]
```

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:32]: I see you have a CMakeLists.txt in the repo. What are all these autotools-deps for? What do we need to patch on windows?

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:39]: Why does this depend on the CPU builds? What would go wrong if someone tried to use the CUDA builds? FTR, I don't think we should constrain this.
```suggestion
    - kaldi
```

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:47]: openfst has a [run-export](https://github.com/conda-forge/openfst-feedstock/blob/main/recipe/meta.yaml#L24), so this gest done by the build. Kaldi doesn't have one yet, but we should (eventually) do that.

Numpy should either have a lower bound or a `*`, otherwise conda smithy will think this recipe needs to be built for different numpy versions (but I don't see it as a host-dependence, so that shouldn't be necessary).

```suggestion
    - kaldi
    - numpy >=1.x
    - librosa
```

**Review by h-vetinari — COMMENTED** [review]
Cool, thanks a lot for the PR & the ping!

Some comments

File: recipes/kalpy/meta.yaml
Comment (mmcauliffe) [reply recipes/kalpy/meta.yaml:10]: Oh this is just a typo left over from a previous one.  Just plain `kalpy` is too close to some other package on pypi, I think `kaipy` is the conflicting one.  So I was struggling with what to name it, but at the moment it's `kalpy-kaldi` which should help with discoverability considering the number of other binding libraries on pip at various stages of abandonment.  In this case, should the conda recipe output both `kalpy` and `kalpy-kaldi`?  Should I figure out another name that allows for more transparency and consistency across both systems?

I'm also on the fence about having it on pypi to begin with, because it does basically require having the shared libraries from the kaldi conda package installed anyway.

The three options I was considering were:

1. (Current set up) Have just sdist on Pypi, installation requires having kaldi shared libraries where KALDI_ROOT is pointing.  Upside is that I can list kalpy in the dependencies for montreal-forced-aligner, but pip installation is not as easy or fully packaged as I would like.  I haven't really looked into the work required to get wheels working for each platform that would not depend on the conda package, but that would simplify pip installation.
2. No package on pypi, just a conda package, which would make more sense in my mind because having a pip package depend on conda packages seems like a broken experience, also doesn't get the autodetection of new versions.
3. Have kalpy be an output of the kaldi feedstock, since they are largely just bindings for the kaldi libraries, but I think this option probably complicates that feedstock too much.

Do you have any thoughts on best practices here?

File: recipes/kalpy/meta.yaml
Comment (mmcauliffe) [reply recipes/kalpy/meta.yaml:35]: Yeah that makes sense, I was getting some errors when I didn't constrain it, but I can try another pass with them now that I've fixed the other blocking errors.

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [reply recipes/kalpy/meta.yaml]: > Do you have any thoughts on best practices here?

I can only really comment on conda-forge, distribution outside of that is up to the respective maintainers. I agree with much of your reasoning (pypi packages depending on conda-forge makes no sense, "better" [for PyPI] would be vendoring libkaldi in the kalpy wheel), but it doesn't really matter for our purposes here.

Also, IIUC `mfa-kaldi` in the readme was a leftover from something else (it's not there anymore in the repo), so that sounds fine as well.

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:20]: `script_env` is not really so much in use. rather than work around bugs there (e.g. if `%LIBRARY_PREFIX%` doesn't work out of the box), it would IMO be easier to just add it to the script section.

```suggestion
  script:
   - export KALDI_ROOT="$PREFIX"        # [unix]
   - set "KALDI_ROOT=%LIBRARY_PREFIX%"  # [win]
    - CXXFLAGS="${CXXFLAGS} -D_LIBCPP_DISABLE_AVAILABILITY" {{ PYTHON }} -m pip install . --no-deps -vv   # [not win]
    - {{ PYTHON }} -m pip install . --no-deps -vv                                                         # [win]
  number: 0
  skip: True  # [py<38]
```

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [reply recipes/kalpy/meta.yaml:37]: It's still there in the host requirements...

**Comment by mmcauliffe** [comment]
@h-vetinari I'm still working on figuring out what's going wrong with windows builds, but linux is giving the same message `ImportError: libcuda.so.1: cannot open shared object file: No such file or directory`.  It doesn't look like it is in the cudatoolkit, so that makes sense, but I'm confused how it gets linked to it in that case.  Do we need to have a cuda build matrix for kalpy like for kaldi to have libcuda.so on the build machine?  Is there something wrong in how Kaldi is linking to the cuda libraries (ref: https://github.com/kaldi-asr/kaldi/blob/master/CMakeLists.txt#L188)?

**Comment by h-vetinari** [comment]
> linux is giving the same message `ImportError: libcuda.so.1: cannot open shared object file: No such file or directory`.

That looks like it needs to load the cuda driver, which we cannot distribute, but which is part of the cuda image that gets used when we build something with `{{ compiler('cuda') }}`. There are IMO two choices:
* use the cpu-version of kaldi in the test phase
* use a "fake" `{{ compiler('cuda') }}` in the build section, so that we use the cuda image by default.

I think the first approach is easier...

**Comment by mmcauliffe** [comment]
@h-vetinari ok looks like everything's passing again if you want to take another look!

File: recipes/kalpy/conda_build_config.yaml
Comment (h-vetinari) [inline recipes/kalpy/conda_build_config.yaml:4]: Delete this file (we've moved to default vs2019 already across all of conda-forge)

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:37]: Numpy is _only_ a run-dependence? Also, you cannot use "1.x" (or, at least, I have no idea how conda would interpret that, it's not a valid version). Conda-forge is already globally building only for numpy>=1.21 anymore, anything below that would not be installed without forcing it anyway.
```suggestion
    - numpy
```

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:36]: Kaldi has a run-export, so you can delete this line here.

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:38]: I find it really confusing that a python package has a name starting with "lib", but whatever...

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:42]: Please add a comment why we're adding this requirement. Also, prefer `=` over semantically relevant whitespace.
```suggestion
  requires:
    # the cuda variant of kaldi requires loading libcuda.so, which we don't have in regular CI
    - kaldi =*=cpu*
```

File: recipes/kalpy/build.sh
Comment (h-vetinari) [inline recipes/kalpy/build.sh:6]: Missing linebreak

File: recipes/kalpy/bld.bat
Comment (h-vetinari) [inline recipes/kalpy/bld.bat:6]: Missing line break

**Review by h-vetinari — COMMENTED** [review]
Looking pretty good, some remaining nits

File: recipes/kalpy/meta.yaml
Comment (h-vetinari) [inline recipes/kalpy/meta.yaml:42]:
```suggestion
    - kaldi =*=cpu*
```

**Review by h-vetinari — APPROVED** [review]
LGTM, thanks a lot!

PTAL @conda-forge/help-python-c

**Comment by h-vetinari** [comment]
@conda-forge/help-python-c, this should be ready (I had a look already, though of course a second review would help, and I don't have the commit bit here), PTAL[^1]!

[^1]: I've noticed that the content of the review-summary often does not cause the same actions as a regular comment, e.g. no xref created, no bot triggered etc.; sorry for the duplicated ping

**Comment by mmcauliffe** [comment]
@conda-forge/help-python-c reminder ping to see if I can get this reviewed and merged.  Thanks!

**Comment by h-vetinari** [comment]
I already approved, but recently, I gained the power to push the button here. :)
This one is good to go.

**Comment by mmcauliffe** [comment]
Awesome, thanks so much!!

**Pull request merged**
